### DISCUSSION TITLE
ftd `type` and `or-type`

### DISCUSSION THREAD

**Discussion by Arpita-Jaiswal** [fastn-stack/fastn#470] [Ideas & RFCs]
For ftd style padding etc which accepts length type values, we are using string as an accepted type because this gives more flexibility, as in, to provide the `percentage`, `px`, `vw`, `vh` etc as value for such style. But since a `string` type can take arbitrary values and be changed on events, there's no check to ensure its validity. So we need a more specific type for such cases and `ftd.length` can help. The `ftd.length` is `or-type` type, narrowing down the cardinality or acceptable values. It can be defined in the following way.

# `$pattern$` proposal

```ftd
-- or-type length:

-- record px:
caption integer value:

-- decimal percent:

-- end: length

-- ftd.text: Hello
padding: 20%
```

There must be some way to define patterns, as in, for percentage, the pattern should be something like `{0}%` or `{}%`.
Consider this

```ftd
-- or-type length:

-- decimal rem:

-- record px:
caption integer value:

-- record percentage:
caption decimal value:
$pattern$: {}%

-- end: length

;; The value here would be of type `length.percentage` as the pattern matches
-- length value: 20%

;; Other way of defining the above could be
-- length.percentage value: 20
```

Refer to https://github.com/ftd-lang/ftd/discussions/314

# Constants And Data

How would we represent the enum:

```rust
pub enum Resizing {
    HugContent,
    FillContainer,
    Fixed(usize)
}
```

```ftd
-- or-type resizing:
-- string constant hug-content: hug-content
-- string constant fill-container: fill-container
-- ftd.length.px constant zero: 0
-- integer constant f: 22
-- ftd.length fixed:
```

When using constants, the name and value can be different for string. How to use them:

```ftd
-- ftd.resizing r: hug-content
-- ftd.resizing.fixed.px rf: 30

-- ftd.row:
horizontal-resizing: hug-content
vertical-resizing.fixed.px: 20
-- end: ftd.row
```

We will not allow:

```ftd
-- ftd.resizing r: resizing.hug-content

-- ftd.row:
horizontal-resizing: resizing.hug-content
-- end: ftd.row
```

# enum

We do not need `enum`.

If all the "clauses" of an or-type were constant and of same type, we could have created:

```ftd
-- string enum resizing:
hug-content: Hug Content
fill-container: Fill Container
```

And usage:

```ftd
-- ftd.row:
horizontal-resizing: resizing.hug-content
-- end: ftd.row
```

Where `ftd.row` is:

```ftd
-- component row:
resizing horizontal-resizing: resizing.hug-content

...

-- end: row
```

# Multiple Matches in PScript

```ftd
-- match $hr:

-- ftd.text: resizing: hug-content
$case$: ftd.resizing.hug-content as _

-- match $len:
$case$: ftd.resizing.fixed as len

    -- ftd.text: resizing percent: { value }
    $case$: ftd.length.percent{value}

    -- ftd.text: horizontal-resizing px: { p.value }
    $case$: ftd.length.px as px

-- end: $len

-- ftd.text: resizing: fill-container
$case$: ftd.resizing.fill-container as _

-- end: $hr
```

`as _` is optional, so least case could have been: `$case$: ftd.resizing.fill-container` as well.

Note that we are doing both aliasing, and "record-destructuring". Destructuring only works in case of record.

We can also do:

```ftd
-- match $hr:

-- ftd.text: resizing: hug-content
$case$: ftd.resizing.hug-content

-- ftd.text: resizing percent: { p.value }
$case$: ftd.resizing.fixed.percent as p

-- ftd.text: horizontal-resizing is in px
$case$: ftd.resizing.fixed.px

-- ftd.text: resizing: fill-container
$case$: ftd.resizing.fill-container

-- end: $hr
```

# Multiple Matches in FScript

```ftd
-- string w(r):
ftd.resizing r:

match(r) {
   ftd.resizing.hug-content as _ {
       "hug content"
   },
   ftd.resizing.fixed as f {
       match(f) {
            ftd.length.px{value} as p {
                  "px: {value}" 
            }
       }
   }
}
```

Also if you do not need the value, you can leave out `as <alias>`.

Or

```ftd
-- string w(r):
ftd.resizing r:

match(r) {
   ftd.resizing.hug-content {
       "hug content"
   },
   ftd.resizing.fixed.px {
         "it is some pixel value" 
   }
}
```

# Using Existing Records Inside Or-Type

```ftd
-- record person:
caption name:
body bio

-- record company:
caption name:
body registered-address:

-- or-type entity:

-- company c:
-- person p:

-- end: entity
```

# ~Using Internal Records AS Type~

```ftd
-- component foo:
length.px p:

-- end: foo

-- foo:
p: 20
```

But if we write

```ftd
-- length.px p:
```

What is the type of `p`?

So we can not do this. Define the record outside the `or-type` and use it in both places.

# subname proposal

Other alternative is we do:

```ftd
-- ftd.text: Hello
padding.percent: 2
```

This is in line with `or-type` initialisation:

```ftd
-- length.percent len: 20
```

We have created a variable called `len` or type `length`, using the `.percent` clause of the or-type.

We can then do:

```ftd
-- ftd.text:
padding: $len
```

How to verify `or-type` variable in FTDScript?

```ftd
-- ftd.text: foo
$on-click$: { len = length.px(value=20) }
```

Rule: We can not read an or-type without a match.

```ftd
-- ftd.text: foo
$on-click$: {
    match (len) {
        length.percent(p) {
            p.value += 1;
        }
        length.px(p) {
            p.value += 20;
        }
    }
}
```

Currently we are going to implement this

```ftd
-- decimal $value: 20
-- length.percent $len: $value

-- ftd.text: Click
padding: $len
$on-click$: { value +=  1 }
```

or 

```ftd
-- decimal $value: 20

-- ftd.text: Click
padding.percent: $value
padding.px if { flag }: 50
$on-click$: { value +=  1 }
```

# P1 Syntax for match

```ftd
-- match $len:

-- ftd.text: percent: { p.value }
$case$: length.percent as p

-- ftd.text: px: { p.value }
$case$: length.px as px

-- end: match
```

While we have used same component, `ftd.text`, in both "branches" we could have used any.

# Initialising a variable based on `or-type` value

```ftd
-- optional decimal f: {
    match (len) {
        length.percent(p) { p.value * 2.0 }
        _ { null }
    }
}
```

# Using check/validation

Refer to https://github.com/ftd-lang/ftd/discussions/313

Another thought is we can use checks for validation of value passed to string and narrow down the cardinality.
Consider this

```ftd
-- bool validate-length(a):
string a:

if(a.is-int()) {
 return true;
}

if(a.ends-with('%') && a[..-1].trim().is-decimal()) {
  return true;
}

return false;


-- string length: 20
$check$: validate-length
```

**Comment by Arpita-Jaiswal** [comment]
```ftd
-- or-type length:

-- record px:
caption integer value:

-- record percent:
caption decimal value:

-- end: length



-- or-type resizing:

-- constant string hug-content: hug-content
-- constant string fill-container: fill-container
-- ftd.length fixed:

-- end: resizing




-- or-type background:

-- ftd.color solid:

-- end: background
```

**Reply by amitu** [reply]
We discussed these names:

- Constant Variant (eg `hug-content` and `fill-container` in your example)
- Regular Variant (eg `fixed` in your example)
- Anonymous Record Variant (eg `px` and `percent` in your example)
- Anonymous OrType Variant

And example of anonymous or-type variant:

```ftd
-- or-type r:

-- or-type f:

    -- integer px:
    -- decimal inches:

-- end: f

-- end: r
```

**Reply by amitu** [reply]
@Arpita-Jaiswal What if we can do this as well:

```ftd
-- or-type resizing:

-- constant string hug-content: Hug Content
-- constant string fill-container: Fill Container
-- ftd.length fixed:

-- end: resizing

-- resizing r: hug-content
-- resizing.fixed.px p: 20
```

Since the type of is `resizing`, we know what `hug-content` means. Less code to write, same type safety.

**Reply by amitu** [reply]
```ftd
-- or-type spacing:

-- constant string in-between: In Between
-- ftd.length fixed:

-- end: spacing
```
